misc: migrate from dep to go modules

invidian opened the pull request:
Closes #345

Note: this should probably be rebased once #343 is merged and then reviewed to make sure CI passes.

invidian committed:
misc: migrate from dep to go modules

Closes #345

Signed-off-by: Mateusz Gozdek <mateusz@kinvolk.io>

invidian committed:
misc: update all modules

Note: github.com/containernetworking/plugins must be v0.7.0 as from
version v0.8.0 they removed code resposible for network namespace
creation.

See
https://github.com/containernetworking/plugins/commit/a0eac8d7d990adc7621e6ac0adf8b351e6e6c34b#diff-d3d859d29b12afe98694dc54a7ced6d9
for details.

Refs #344

Signed-off-by: Mateusz Gozdek <mateusz@kinvolk.io>

invidian commented:
Oops, it seems that Fedora 28, which is used by default ships go 1.10, which don't have modules support. I'll update defaults to use Fedora 30 and check if everything works as expected still.

dongsupark commented:
I've tested on my local system, Fedora 30. It fails with:

```
$ DOCKERIZED=n make
go mod download
go build -ldflags "-X main.version=v0.3.0-37-g24499024" \
        ./cmd/kube-spawn
# github.com/kinvolk/kube-spawn/pkg/cnispawn
pkg/cnispawn/netns.go:38:16: undefined: ns.NewNS
make: *** [Makefile:26: all] Error 2
```

I'm not completely sure, but with CNI plugins 0.7 or newer, we cannot call `NewNS` which does not exist, right?
No wonder, because kube-spawn has never been tested with newer CNI plugins like 0.7.

invidian commented:
@dongsupark see my commit message:
```
Note: github.com/containernetworking/plugins must be v0.7.0 as from
version v0.8.0 they removed code resposible for network namespace
creation.

See
containernetworking/plugins@a0eac8d#diff-d3d859d29b12afe98694dc54a7ced6d9
for details.
```

It also builds in Vagrant, so it must be your local environment issue. I suggest executing `GO111MODULE=on go clean -modcache`.

In any case, also Vagrantfile and scripts require changing to use go modules, so this PR is incomplete.

dongsupark commented on Makefile:
```diff
@@ -5,8 +5,9 @@ PREFIX ?= /usr
 BINDIR ?= ${PREFIX}/bin
 UID=$(shell id -u)
 GOCACHEDIR=$(shell which go >/dev/null 2>&1 && go env GOCACHE || echo "$(HOME)/.cache/go-build")
+GO111MODULE=on
```
I think the issue is that the line above does not set the env variable `$GO111MODULE` expected.

So if I manually set `GO111MODULE=on` for the go build command, it works fine.
On the other hand, if I run it without `GO111MODULE=on`, it sometimes fails, because it tries to look for the cni plugins repo under my GOPATH. If my local cni plugins repo is based on 0.7, it works fine. If not, it fails.
That's not exactly what users would expect to happen.

invidian replied:
Ah, that's my missing knowledge regarding `Makefile`. Should I add `export GO111MODULE=on` to `all` and `update-vendor` then?

dongsupark replied:
Yes, you can do so.
Or you can simply add `GO111MODULE=on` only for go build commands.
If it works, I don't care much.

invidian commented:
I pushed fixed `Makefile`. However, there are still some changed pending, which needs to be pushed (updating Fedora release etc).

invidian committed:
scripts/vagrant-mod-env.sh: allow subsequent provisioner run

Ignore when systemd-resolved is already disabled - for troubleshooting
when running provisioning mutliple times, is-enabled will exit 1, which
aborts provisioning. It should be safe to simply skip it then with.

Signed-off-by: Mateusz Gozdek <mateusz@kinvolk.io>

invidian committed:
scripts/vagrant-setup-env.sh: install CNI plugins with explicit GO111MODULE

Fetch CNI plugins without using go modules - there seems to be an issue
when using Go modules, so to reduce amount of changes done, simply use
GOPATH as before.

Signed-off-by: Mateusz Gozdek <mateusz@kinvolk.io>

invidian committed:
Vagrantfile: update to Fedora 30

Fedora 28 ships Go version 1.10 which does not support modules, so the
correct way of bumping this dependency is to switch to newer Fedora
release.

Signed-off-by: Mateusz Gozdek <mateusz@kinvolk.io>

dongsupark commented on Makefile:
```diff
@@ -15,21 +15,22 @@ all:
+	export GO111MODULE=on
+	go mod download
+	go build -ldflags "-X main.version=$(VERSION)" \
 		./cmd/kube-spawn
```
Sorry, I was not specific.
The current Makefile still does not set `GO111MODULE` correctly.

In Makefile, you need to specify env variables either in the same line you want to run, or in one of the multiple lines separated by `;`.

For example you can do either

```
        go mod download
        GO111MODULE=on go build
```

or

```
        export GO111MODULE=on; \
        go mod download; \
        go build
```
See also https://stackoverflow.com/questions/42993169/setting-environment-variable-in-makefile-and-reading-it-from-python-script-does

invidian replied:
Yes, I have this one fixed already, just didn't push it yet. Testing takes time...

dongsupark commented on Makefile:
```diff
@@ -15,21 +15,22 @@ all:
+update-vendor:
+	export GO111MODULE=on
+	go get -u
+	go mod tidy
```
Ditto

dongsupark reviewed (commented):
See inline comments:

invidian committed:
vagrant-all.sh: allow specifying extra vagrant flags

If provisioning fails, one can add KUBESPAWN_VAGRANT_EXTRA_FLAGS=--provision
to start provisioning over, which saves time during testing.

Signed-off-by: Mateusz Gozdek <mateusz@kinvolk.io>

invidian committed:
Vagrantfile: update Ubuntu to 19.04

This Ubuntu version has go 1.12 version in main repositories.
We require go version 1.11 as we use go modules.

Signed-off-by: Mateusz Gozdek <mateusz@kinvolk.io>

invidian committed:
scripts/vagrant-setup-env.sh: add Ubuntu go bnary path to build.sh

Ubuntu 19.04 ships Go version 1.10 by default, however we install
version 1.12 because of Go modules and this version is not installed
in default PATH, so we need to add it manually.

Signed-off-by: Mateusz Gozdek <mateusz@kinvolk.io>

invidian committed:
scripts/vagrant-setup-env.sh: download kubectl only if missing

If provisioning is executed again, it won't download kubectl twice.

Signed-off-by: Mateusz Gozdek <mateusz@kinvolk.io>

invidian committed:
scripts/vagrant-setup-env.sh: create cluster only once

If provisioning script fails and is executed again, don't try to
create cluster again and simply use existing files.

Signed-off-by: Mateusz Gozdek <mateusz@kinvolk.io>

invidian committed:
scripts/vagrant-setup-env.sh: remove machinectl download artifacts

After downloading, machinectl creates another image, which takes up
space, so we should clean it up.

Also, if pull-raw fails to download image, it does not return error and
then kube-spawn tries to do it as well, so we should try to cleanup
after it's run as well.

Signed-off-by: Mateusz Gozdek <mateusz@kinvolk.io>

invidian committed:
scripts/vagrant-setup-env.sh: set -o pipefail

So if some opeartion fails, we don't ignore it.

Signed-off-by: Mateusz Gozdek <mateusz@kinvolk.io>

invidian committed:
pkg: improve support for systemd >= 240 and native btrfs parititions

Systemd prior to version 240 was creating /var/lib/machines.raw as btrfs
volume for storing machined disks. However, this functionality is now
remove and now it stores it as regular files. More details here:
https://github.com/systemd/systemd/blob/401faa3533280b05fee972e0c64885caf4b31e4c/units/var-lib-machines.mount#L10

Additionally, it someone already have /var/lib/machines baked by btrfs,
machined won't create machines.raw, which will also cause setting up
cluster to fail.

This patch adds check, if machineImage exists and it not, it fallbacks
to just checking disk space available for /var/lib/machines.

Refs #339 #293

Signed-off-by: Mateusz Gozdek <mateusz@kinvolk.io>

invidian committed:
gitignore: ignore .ssh_config

This seems to be Vagrant artifact, which we don't care about.

dongsupark reviewed (approved):
LGTM in general.

First of all, `go modules` works well as expected. Great!

Vagrant test scripts work as well, both for Fedora and Ubuntu.

Though I'm still seeing the issue of `kube-spawn start` hanging, so I'm not sure this PR would fix https://github.com/kinvolk/kube-spawn/issues/339. Anyway I'm fine with merging this PR, as that looks like a separate issue.

And in general I would like to see multiple separate PRs, instead of 16 commits in a single PR. Having an individual PR for each topic would make reviews a lot easier.

Minor issues inline: (Though I don't have strong opinions)

invidian commented on Vagrantfile:
```diff
@@ -45,10 +45,10 @@ Vagrant.configure("2") do |config|
-    config.vm.box = "generic/ubuntu1804"
-    config.vm.provision "shell", inline: "apt-get update; DEBIAN_FRONTEND=noninteractive apt-get install -y btrfs-progs git golang iptables make policykit-1 qemu-utils rinetd selinux-utils systemd-container"
+    config.vm.box = "generic/ubuntu1904"
+    config.vm.provision "shell", inline: "apt-get update; DEBIAN_FRONTEND=noninteractive apt-get install -y btrfs-progs git golang-1.11 iptables make policykit-1 qemu-utils rinetd selinux-utils systemd-container"
```
I picked 1.11 as it's the same version shipped by Fedora 29 and Debian testing. I upgraded to Fedora 30 first, then rolled back to Fedora 29 to see if that solves cluster spawning issues, then updated to Fedora 30 again. We can bump this to 1.12.

invidian committed:
vagrant-all.sh: add set -u

To protect from unbound variables.

Signed-off-by: Mateusz Gozdek <mateusz@kinvolk.io>

invidian commented on vagrant-all.sh:
```diff
@@ -8,7 +8,7 @@ export KUBESPAWN_REDIRECT_TRAFFIC="true"
 KUBESPAWN_PROVIDER=${KUBESPAWN_PROVIDER:-virtualbox}
 
-vagrant up $KUBESPAWN_DISTRO --provider=$KUBESPAWN_PROVIDER
+vagrant up $KUBESPAWN_DISTRO --provider=$KUBESPAWN_PROVIDER ${KUBESPAWN_VAGRANT_EXTRA_FLAGS}
```
We don't because we don't do `set -u`. However, it's a good idea to add both (declaration and -u), so I'll do it.

invidian commented:
Thanks for the review and feedback @dongsupark.

>And in general I would like to see multiple separate PRs, instead of 16 commits in a single PR. Having an individual PR for each topic would make reviews a lot easier.

I know amount of commits started piling up and I'm not the fan of it either. Unfortunately the more I was testing, the more possible improvements I was fining and implementing. I can still split it into separated PRs if you like.

>Though I'm still seeing the issue of kube-spawn start hanging, so I'm not sure this PR would fix #339. Anyway I'm fine with merging this PR, as that looks like a separate issue.

You are right about this. By the time of writing, I assumed this was the only issue with systemd >= 240, turns out there are some other issues with networking, which I didn't have time to investigate yet. I changed commit message to only reference to these issues.

I addressed your comments, so PR is ready for another round of review.

dongsupark commented:
Looks good. Thanks!
